File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseTimePeriodExtractor.cs
﻿// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

using Microsoft.Recognizers.Text.InternalCache;
using Microsoft.Recognizers.Text.Utilities;
using DateObject = System.DateTime;

namespace Microsoft.Recognizers.Text.DateTime
{
    public class BaseTimePeriodExtractor : IDateTimeExtractor
    {
        public static readonly string ExtractorName = Constants.SYS_DATETIME_TIMEPERIOD; // "TimePeriod";

        private static readonly ResultsCache<ExtractResult> ResultsCache = new ResultsCache<ExtractResult>();

        private readonly ITimePeriodExtractorConfiguration config;

        private readonly string keyPrefix;

        public BaseTimePeriodExtractor(ITimePeriodExtractorConfiguration config)
        {
            this.config = config;
            keyPrefix = string.Intern(config.Options + "_" + config.LanguageMarker);
        }

        public List<ExtractResult> Extract(string text)
        {
            return Extract(text, DateObject.Now);
        }

        public List<ExtractResult> Extract(string text, DateObject reference)
        {
            List<ExtractResult> results;

            if ((this.config.Options & DateTimeOptions.NoProtoCache) != 0)
            {
                results = ExtractImpl(text, reference);
            }
            else
            {
                var key = (keyPrefix, text, reference);

                results = ResultsCache.GetOrCreate(key, () => ExtractImpl(text, reference));
            }

            return results;
        }

        private List<ExtractResult> ExtractImpl(string text, DateObject reference)
        {
            var tokens = new List<Token>();
            tokens.AddRange(MatchSimpleCases(text));
            tokens.AddRange(MergeTwoTimePoints(text, reference));
            tokens.AddRange(MatchTimeOfDay(text));

            // Handle pure number cases like "from 6 to 7" cannot be extracted as time ranges under Calendar Mode
            if ((this.config.Options & DateTimeOptions.CalendarMode) != 0)
            {
                tokens.AddRange(MatchPureNumberCases(text));
            }

            var timePeriodErs = Token.MergeAllTokens(tokens, text, ExtractorName);

            if ((this.config.Options & DateTimeOptions.EnablePreview) != 0)
            {
                timePeriodErs = TimeZoneUtility.MergeTimeZones(timePeriodErs, config.TimeZoneExtractor.Extract(text, reference), text);
            }

            // Filter ambiguous extractions e.g. 'morgen' in German and Dutch
            timePeriodErs = this.config.ApplyPotentialPeriodAmbiguityHotfix(text, timePeriodErs);

            return timePeriodErs;
        }

        // Cases like "from 3 to 5am" or "between 3:30 and 5" are extracted here
        // Note that cases like "from 3 to 5" will not be extracted here because no "am/pm" or "hh:mm" to infer it's a time period
        // Also cases like "from 3:30 to 4 people" shuold not be extracted as a time period
        private List<Token> MatchSimpleCases(string text)
        {
            var ret = new List<Token>();
            foreach (var regex in this.config.SimpleCasesRegex)
            {
                var matches = regex.Matches(text);

                foreach (Match match in matches)
                {
                    // Cases like "from 10:30 to 11", don't necessarily need "am/pm"
                    if (match.Groups[Constants.MinuteGroupName].Success || match.Groups[Constants.SecondGroupName].Success)
                    {
                        // Cases like "from 3:30 to 4" should be supported
                        // Cases like "from 3:30 to 4 on 1/1/2015" should be supported
                        // Cases like "from 3:30 to 4 people" is considered not valid
                        bool endWithValidToken = false;

                        // "No extra tokens after the time period"
                        if (match.Index + match.Length == text.Length)
                        {
                            endWithValidToken = true;
                        }
                        else
                        {
                            var afterStr = text.Substring(match.Index + match.Length);

                            // "End with general ending tokens or "TokenBeforeDate" (like "on")
                            var endWithGeneralEndings = this.config.GeneralEndingRegex.Match(afterStr).Success;
                            var endWithAmPm = match.Groups[Constants.RightAmPmGroupName].Success;

                            if (endWithGeneralEndings || endWithAmPm || afterStr.TrimStart().StartsWith(this.config.TokenBeforeDate, StringComparison.Ordinal))
                            {
                                endWithValidToken = true;
                            }
                            else if ((this.config.Options & DateTimeOptions.EnablePreview) != 0)
                            {
                                endWithValidToken = StartsWithTimeZone(afterStr);
                            }
                        }

                        if (endWithValidToken)
                        {
                            ret.Add(new Token(match.Index, match.Index + match.Length));
                        }
                    }
                    else
                    {
                        // Is there "pm" or "am"?
                        var matchPmStr = match.Groups[Constants.PmGroupName].Value;
                        var matchAmStr = match.Groups[Constants.AmGroupName].Value;
                        var descStr = match.Groups[Constants.DescGroupName].Value;

                        // Check "pm", "am"
                        if (!string.IsNullOrEmpty(matchPmStr) || !string.IsNullOrEmpty(matchAmStr) || !string.IsNullOrEmpty(descStr))
                        {
                            ret.Add(new Token(match.Index, match.Index + match.Length));
                        }
                        else
                        {
                            var afterStr = text.Substring(match.Index + match.Length);

                            if ((this.config.Options & DateTimeOptions.EnablePreview) != 0 && StartsWithTimeZone(afterStr))
                            {
                                ret.Add(new Token(match.Index, match.Index + match.Length));
                            }
                        }
                    }
                }
            }

            return ret;
        }

        private bool StartsWithTimeZone(string afterText)
        {
            var startsWithTimeZone = false;

            var timeZoneErs = config.TimeZoneExtractor.Extract(afterText);
            var firstTimeZone = timeZoneErs.OrderBy(t => t.Start).FirstOrDefault();

            if (firstTimeZone != null)
            {
                var beforeText = afterText.Substring(0, firstTimeZone.Start ?? 0);

                if (string.IsNullOrWhiteSpace(beforeText))
                {
                    startsWithTimeZone = true;
                }
            }

            return startsWithTimeZone;
        }

        private List<Token> MergeTwoTimePoints(string text, DateObject reference)
        {
            var ret = new List<Token>();
            var ers = this.config.SingleTimeExtractor.Extract(text, reference);

            // Merge "{TimePoint} to {TimePoint}", "between {TimePoint} and {TimePoint}"

            // Handling ending number as a time point.
            var numErs = this.config.IntegerExtractor.Extract(text);

            // Check if it is an ending number
            if (numErs.Count > 0)
            {
                var timeNumbers = new List<ExtractResult>();

                // check if it is a ending number
                var endingNumber = false;
                var num = numErs[numErs.Count - 1];
                if (num.Start + num.Length == text.Length)
                {
                    endingNumber = true;
                }
                else
                {
                    var afterStr = text.Substring(num.Start + num.Length ?? 0);
                    var endingMatch = this.config.GeneralEndingRegex.Match(afterStr);
                    if (endingMatch.Success)
                    {
                        endingNumber = true;
                    }
                }

                if (endingNumber)
                {
                    timeNumbers.Add(num);
                }

                var i = 0;
                var j = 0;
                while (i < numErs.Count)
                {
                    // find subsequent time point
                    var numEndPoint = numErs[i].Start + numErs[i].Length;
                    while (j < ers.Count && ers[j].Start <= numEndPoint)
                    {
                        j++;
                    }

                    if (j >= ers.Count)
                    {
                        break;
                    }

                    // check connector string
                    var midStr = text.Substring(numEndPoint ?? 0, ers[j].Start - numEndPoint ?? 0);

                    if (config.TillRegex.IsExactMatch(midStr, trim: true) || config.IsConnectorToken(midStr.Trim()))
                    {
                        timeNumbers.Add(numErs[i]);
                    }

                    i++;
                }

                // check overlap
                foreach (var timeNum in timeNumbers)
                {
                    var overlap = false;
                    foreach (var er in ers)
                    {
                        if (er.Start <= timeNum.Start && er.Start + er.Length >= timeNum.Start)
                        {
                            overlap = true;
                        }
                    }

                    if (!overlap)
                    {
                        ers.Add(timeNum);
                    }
                }

                ers.Sort((x, y) => x.Start - y.Start ?? 0);
            }

            var idx = 0;
            while (idx < ers.Count - 1)
            {
                var middleBegin = ers[idx].Start + ers[idx].Length ?? 0;
                var middleEnd = ers[idx + 1].Start ?? 0;

                if (middleEnd - middleBegin <= 0)
                {
                    idx++;
                    continue;
                }

                var middleStr = text.Substring(middleBegin, middleEnd - middleBegin).Trim();

                // Handle "{TimePoint} to {TimePoint}"
                if (config.TillRegex.IsExactMatch(middleStr, trim: true))
                {
                    var periodBegin = ers[idx].Start ?? 0;
                    var periodEnd = (ers[idx + 1].Start ?? 0) + (ers[idx + 1].Length ?? 0);

                    // Handle "from"
                    var beforeStr = text.Substring(0, periodBegin).TrimEnd();
                    var afterStr = text.Substring(periodEnd, text.Length - periodEnd);
                    if (this.config.GetFromTokenIndex(beforeStr, out var fromIndex))
                    {
                        // Handle "from"
                        periodBegin = fromIndex;
                    }
                    else if (this.config.GetBetweenTokenIndex(beforeStr, out var betweenIndex))
                    {
                        // Handle "between"
                        periodBegin = betweenIndex;
                    }
                    else if (this.config.CheckBothBeforeAfter && this.config.GetBetweenTokenIndex(afterStr, out var afterIndex))
                    {
                        // Handle "between" in afterStr
                        periodEnd += afterIndex;
                    }

                    ret.Add(new Token(periodBegin, periodEnd));
                    idx += 2;
                    continue;
                }

                // Handle "between {TimePoint} and {TimePoint}"
                if (this.config.IsConnectorToken(middleStr))
                {
                    var periodBegin = ers[idx].Start ?? 0;
                    var periodEnd = (ers[idx + 1].Start ?? 0) + (ers[idx + 1].Length ?? 0);

                    // Handle "between"
                    var beforeStr = text.Substring(0, periodBegin).Trim();
                    if (this.config.GetBetweenTokenIndex(beforeStr, out int betweenIndex))
                    {
                        periodBegin = betweenIndex;
                        ret.Add(new Token(periodBegin, periodEnd));
                        idx += 2;
                        continue;
                    }

                    // handle "between...and..." case when "between" follows the datepoints
                    var afterStr = text.Substring(periodEnd, text.Length - periodEnd);
                    if (this.config.CheckBothBeforeAfter && this.config.GetBetweenTokenIndex(afterStr, out int afterIndex))
                    {
                        periodEnd += afterIndex;
                        ret.Add(new Token(periodBegin, periodEnd));

                        // merge two tokens here, increase the index by two
                        idx += 2;
                        continue;
                    }
                }

                idx++;
            }

            return ret;
        }

        private List<Token> MatchTimeOfDay(string text)
        {
            var ret = new List<Token>();
            var matches = this.config.TimeOfDayRegex.Matches(text);
            foreach (Match match in matches)
            {
                Metadata metadata = null;

                if (match.Groups[Constants.MealTimeGroupName].Success)
                {
                    metadata = new Metadata
                    {
                        IsMealtime = true,
                    };
                }

                var token = new Token(match.Index, match.Index + match.Length, metadata);

                ret.Add(token);
            }

            return ret;
        }

        // Support cases like "from 6 to 7" which are pure number ranges
        // Only when the number range is at the end of a sentence, it will be considered as a time range
        private List<Token> MatchPureNumberCases(string text)
        {
            var ret = new List<Token>();
            foreach (var regex in this.config.PureNumberRegex)
            {
                var matches = regex.Matches(text);
                foreach (Match match in matches)
                {
                    var afterStr = text.Substring(match.Index + match.Length);
                    var endingMatch = this.config.GeneralEndingRegex.Match(afterStr);
                    if (endingMatch.Success)
                    {
                        ret.Add(new Token(match.Index, match.Index + match.Length));
                    }
                }
            }

            return ret;
        }
    }
}

File: Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/extractors/BaseTimePeriodExtractor.java
package com.microsoft.recognizers.text.datetime.extractors;

import com.microsoft.recognizers.text.ExtractResult;
import com.microsoft.recognizers.text.datetime.Constants;
import com.microsoft.recognizers.text.datetime.DateTimeOptions;
import com.microsoft.recognizers.text.datetime.extractors.config.ITimePeriodExtractorConfiguration;
import com.microsoft.recognizers.text.datetime.extractors.config.ResultIndex;
import com.microsoft.recognizers.text.datetime.utilities.ConditionalMatch;
import com.microsoft.recognizers.text.datetime.utilities.RegexExtension;
import com.microsoft.recognizers.text.datetime.utilities.TimeZoneUtility;
import com.microsoft.recognizers.text.datetime.utilities.Token;
import com.microsoft.recognizers.text.utilities.Match;
import com.microsoft.recognizers.text.utilities.RegExpUtility;
import com.microsoft.recognizers.text.utilities.StringUtility;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import java.util.regex.Pattern;

public class BaseTimePeriodExtractor implements IDateTimeExtractor {

    private final ITimePeriodExtractorConfiguration config;

    @Override
    public String getExtractorName() {
        return Constants.SYS_DATETIME_TIMEPERIOD;
    }

    public BaseTimePeriodExtractor(ITimePeriodExtractorConfiguration config) {
        this.config = config;
    }

    @Override
    public List<ExtractResult> extract(String input, LocalDateTime reference) {

        List<Token> tokens = new ArrayList<>();
        tokens.addAll(matchSimpleCases(input));
        tokens.addAll(mergeTwoTimePoints(input, reference));
        tokens.addAll(matchTimeOfDay(input));

        List<ExtractResult> timePeriodErs = Token.mergeAllTokens(tokens, input, getExtractorName());

        if (config.getOptions().match(DateTimeOptions.EnablePreview)) {
            timePeriodErs = TimeZoneUtility.mergeTimeZones(timePeriodErs, config.getTimeZoneExtractor().extract(input, reference), input);
        }

        return timePeriodErs;
    }

    @Override
    public List<ExtractResult> extract(String input) {
        return this.extract(input, LocalDateTime.now());
    }

    // Cases like "from 3 to 5am" or "between 3:30 and 5" are extracted here
    // Note that cases like "from 3 to 5" will not be extracted here because no "am/pm" or "hh:mm" to infer it's a time period
    // Also cases like "from 3:30 to 4 people" should not be extracted as a time period
    private List<Token> matchSimpleCases(String input) {

        List<Token> ret = new ArrayList<>();

        for (Pattern regex : this.config.getSimpleCasesRegex()) {
            Match[] matches = RegExpUtility.getMatches(regex, input);

            for (Match match : matches) {
                // Cases like "from 10:30 to 11", don't necessarily need "am/pm"
                if (!match.getGroup(Constants.MinuteGroupName).value.equals("") || !match.getGroup(Constants.SecondGroupName).value.equals("")) {
                    // Cases like "from 3:30 to 4" should be supported
                    // Cases like "from 3:30 to 4 on 1/1/2015" should be supported
                    // Cases like "from 3:30 to 4 people" is considered not valid
                    Boolean endWithValidToken = false;

                    // "No extra tokens after the time period"
                    if (match.index + match.length == input.length()) {
                        endWithValidToken = true;
                    } else {
                        String afterStr = input.substring(match.index + match.length);

                        // "End with general ending tokens or "TokenBeforeDate" (like "on")
                        boolean endWithGeneralEndings = Arrays.stream(RegExpUtility.getMatches(this.config.getGeneralEndingRegex(), afterStr))
                                .findFirst().isPresent();
                        boolean endWithAmPm = !match.getGroup(Constants.RightAmPmGroupName).value.equals("");
                        if (endWithGeneralEndings || endWithAmPm || afterStr.trim().startsWith(this.config.getTokenBeforeDate())) {
                            endWithValidToken = true;
                        } else if (this.config.getOptions().match(DateTimeOptions.EnablePreview)) {
                            endWithValidToken = startsWithTimeZone(afterStr);
                        }
                    }

                    if (endWithValidToken) {
                        ret.add(new Token(match.index, match.index + match.length));
                    }
                } else {
                    // Is there Constants.PmGroupName or Constants.AmGroupName ?
                    String pmStr = match.getGroup(Constants.PmGroupName).value;
                    String amStr = match.getGroup(Constants.AmGroupName).value;
                    String descStr = match.getGroup(Constants.DescGroupName).value;

                    // Check "pm", "am"
                    if (!StringUtility.isNullOrEmpty(pmStr) || !StringUtility.isNullOrEmpty(amStr) || !StringUtility.isNullOrEmpty(descStr)) {
                        ret.add(new Token(match.index, match.index + match.length));
                    } else {
                        String afterStr = input.substring(match.index + match.length);

                        if ((this.config.getOptions().match(DateTimeOptions.EnablePreview)) && startsWithTimeZone(afterStr)) {
                            ret.add(new Token(match.index, match.index + match.length));
                        }
                    }
                }
            }
        }

        return ret;
    }

    private boolean startsWithTimeZone(String afterText) {
        boolean startsWithTimeZone = false;

        List<ExtractResult> timeZoneErs = config.getTimeZoneExtractor().extract(afterText);
        Optional<ExtractResult> firstTimeZone = timeZoneErs.stream().sorted(Comparator.comparingInt(t -> t.getStart())).findFirst();

        if (firstTimeZone.isPresent()) {
            String beforeText = afterText.substring(0, firstTimeZone.get().getStart());

            if (StringUtility.isNullOrWhiteSpace(beforeText)) {
                startsWithTimeZone = true;
            }
        }

        return startsWithTimeZone;
    }

    private List<Token> mergeTwoTimePoints(String input, LocalDateTime reference) {

        List<Token> ret = new ArrayList<>();
        List<ExtractResult> ers = this.config.getSingleTimeExtractor().extract(input, reference);

        // Handling ending number as a time point.
        List<ExtractResult> numErs = this.config.getIntegerExtractor().extract(input);

        // Check if it is an ending number
        if (numErs.size() > 0) {
            List<ExtractResult> timeNumbers = new ArrayList<>();

            // check if it is a ending number
            boolean endingNumber = false;
            ExtractResult num = numErs.get(numErs.size() - 1);
            if (num.getStart() + num.getLength() == input.length()) {
                endingNumber = true;
            } else {
                String afterStr = input.substring(num.getStart() + num.getLength());
                Pattern generalEndingRegex = this.config.getGeneralEndingRegex();
                Optional<Match> endingMatch = Arrays.stream(RegExpUtility.getMatches(generalEndingRegex, input)).findFirst();
                if (endingMatch.isPresent()) {
                    endingNumber = true;
                }
            }
            if (endingNumber) {
                timeNumbers.add(num);
            }

            int i = 0;
            int j = 0;

            while (i < numErs.size()) {
                // find subsequent time point
                int numEndPoint = numErs.get(i).getStart() + numErs.get(i).getLength();
                while (j < ers.size() && ers.get(j).getStart() <= numEndPoint) {
                    j++;
                }

                if (j >= ers.size()) {
                    break;
                }

                // check connector string
                String midStr = input.substring(numEndPoint, ers.get(j).getStart());
                Pattern tillRegex = this.config.getTillRegex();
                if (RegexExtension.isExactMatch(tillRegex, midStr, true) || config.hasConnectorToken(midStr.trim())) {
                    timeNumbers.add(numErs.get(i));
                }

                i++;
            }

            // check overlap
            for (ExtractResult timeNum : timeNumbers) {
                boolean overlap = false;
                for (ExtractResult er : ers) {
                    if (er.getStart() <= timeNum.getStart() && er.getStart() + er.getLength() >= timeNum.getStart()) {
                        overlap = true;
                    }
                }

                if (!overlap) {
                    ers.add(timeNum);
                }
            }

            ers.sort((x, y) -> x.getStart() - y.getStart());
        }

        int idx = 0;
        while (idx < ers.size() - 1) {
            int middleBegin = ers.get(idx).getStart() + ers.get(idx).getLength();
            int middleEnd = ers.get(idx + 1).getStart();

            if (middleEnd - middleBegin <= 0) {
                idx++;
                continue;
            }

            String middleStr = input.substring(middleBegin, middleEnd).trim().toLowerCase(java.util.Locale.ROOT);
            Pattern tillRegex = this.config.getTillRegex();

            // Handle "{TimePoint} to {TimePoint}"
            if (RegexExtension.isExactMatch(tillRegex, middleStr, true)) {
                int periodBegin = ers.get(idx).getStart();
                int periodEnd = ers.get(idx + 1).getStart() + ers.get(idx + 1).getLength();

                // Handle "from"
                String beforeStr = StringUtility.trimEnd(input.substring(0, periodBegin)).toLowerCase();
                ResultIndex fromIndex = this.config.getFromTokenIndex(beforeStr);
                ResultIndex betweenIndex = this.config.getBetweenTokenIndex(beforeStr);
                if (fromIndex.getResult()) {
                    // Handle "from"
                    periodBegin = fromIndex.getIndex();
                } else if (betweenIndex.getResult()) {
                    // Handle "between"
                    periodBegin = betweenIndex.getIndex();
                }

                ret.add(new Token(periodBegin, periodEnd));
                idx += 2;
                continue;
            }

            // Handle "between {TimePoint} and {TimePoint}"
            if (this.config.hasConnectorToken(middleStr)) {
                int periodBegin = ers.get(idx).getStart();
                int periodEnd = ers.get(idx + 1).getStart() + ers.get(idx + 1).getLength();

                // Handle "between"
                String beforeStr = input.substring(0, periodBegin).trim().toLowerCase(java.util.Locale.ROOT);
                ResultIndex betweenIndex = this.config.getBetweenTokenIndex(beforeStr);
                if (betweenIndex.getResult()) {
                    periodBegin = betweenIndex.getIndex();
                    ret.add(new Token(periodBegin, periodEnd));
                    idx += 2;
                    continue;
                }
            }

            idx++;
        }

        return ret;
    }

    private List<Token> matchTimeOfDay(String input) {

        List<Token> ret = new ArrayList<>();
        Pattern timeOfDayRegex = this.config.getTimeOfDayRegex();
        Match[] matches = RegExpUtility.getMatches(timeOfDayRegex, input);
        for (Match match : matches) {
            ret.add(new Token(match.index, match.index + match.length));
        }

        return ret;
    }
}
